File: cf/pipeline/bili_refresh.py
"""Bilibili web cookie refresh, executed inside the container so the HTTP calls
egress through WARP (PROXY_URLS) instead of the Worker's Cloudflare datacenter
IP — Bilibili risk-controls the passport.bilibili.com endpoints from datacenter
IPs (HTTP 412), so the refresh must run from a WARP exit.

The Worker (lib/biliRefresh.ts) POSTs {cookie, refresh_token, force} to the
/refresh-cookie endpoint and persists the returned {cookie, refresh_token} back
to KV. Flow per the bilibili-API-collect login docs:
  1. GET  /x/passport-login/web/cookie/info?csrf=<bili_jct> -> {refresh, timestamp}
  2. correspondPath = hex(RSA-OAEP-SHA256(pubkey, "refresh_"+timestamp))
     GET https://www.bilibili.com/correspond/1/<path>       -> #1-name = refresh_csrf
  3. POST /x/passport-login/web/cookie/refresh              -> new Set-Cookie + new refresh_token
  4. POST /x/passport-login/web/confirm/refresh (new csrf, OLD token)
"""

from __future__ import annotations

import binascii
import logging
import os
import re

import httpx
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import padding

logger = logging.getLogger("pipeline")

UA = (
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 "
    "(KHTML, like Gecko) Chrome/124.0 Safari/537.36"
)

# Bilibili's well-known RSA public key for correspondPath generation.
_PUBKEY_PEM = b"""-----BEGIN PUBLIC KEY-----
MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQDLgd2OAkcGVtoE3ThUREbio0Eg
Uc/prcajMKXvkCKFCWhJYJcLkcM2DKKcSeFpD/j6Boy538YXnR6VhcuUJOhH2x71
nzPjfdTcqMz7djHum0qSZA0AyCBDABUqCrfNgCiJ00Ra7GmRj+YCK1NJEuewlb40
JNrRuoEUXpabUzGB8QIDAQAB
-----END PUBLIC KEY-----"""


def _correspond_path(timestamp: int) -> str:
    pub = serialization.load_pem_public_key(_PUBKEY_PEM)
    ciphertext = pub.encrypt(
        f"refresh_{timestamp}".encode(),
        padding.OAEP(mgf=padding.MGF1(hashes.SHA256()), algorithm=hashes.SHA256(), label=None),
    )
    return binascii.hexlify(ciphertext).decode()


def _proxy_candidates() -> list[str | None]:
    """Same PROXY_URLS contract as the yt-dlp adapter: WARP SOCKS5 proxies first,
    then `direct`. Refresh rotates through them until one isn't risk-controlled."""
    raw = os.environ.get("PROXY_URLS", "").strip()
    if not raw:
        return [None]
    out: list[str | None] = []
    for part in raw.split(","):
        p = part.strip()
        if p:
            out.append(None if p.lower() == "direct" else p)
    return out or [None]


def _parse_cookie(s: str) -> dict[str, str]:
    jar: dict[str, str] = {}
    for part in (s or "").split(";"):
        t = part.strip()
        if "=" in t:
            k, v = t.split("=", 1)
            if k.strip():
                jar[k.strip()] = v.strip()
    return jar


def _serialize(jar: dict[str, str]) -> str:
    return "; ".join(f"{k}={v}" for k, v in jar.items())


def refresh(cookie: str, refresh_token: str, force: bool = False) -> dict:
    """Refresh the cookie if Bilibili says it needs it (or force). Returns
    {refreshed, reason, cookie?, refresh_token?}. Never raises — errors are
    reported in `reason` so the Worker keeps the previous KV value."""
    if not cookie:
        return {"refreshed": False, "reason": "no cookie"}
    if not refresh_token:
        return {"refreshed": False, "reason": "no refresh_token"}
    jar = _parse_cookie(cookie)
    if not jar.get("bili_jct"):
        return {"refreshed": False, "reason": "cookie missing bili_jct"}

    last_err: Exception | None = None
    for proxy in _proxy_candidates():
        try:
            return _refresh_once(dict(jar), jar["bili_jct"], refresh_token, force, proxy)
        except Exception as exc:  # noqa: BLE001 — rotate to the next proxy
            last_err = exc
            logger.warning("bili refresh via proxy=%s failed: %s", proxy or "direct", exc)
            continue
    return {"refreshed": False, "reason": f"{type(last_err).__name__}: {last_err}"}


def _refresh_once(jar: dict[str, str], csrf: str, refresh_token: str, force: bool, proxy: str | None) -> dict:
    cookie_str = _serialize(jar)
    with httpx.Client(proxy=proxy, headers={"user-agent": UA}, timeout=30, follow_redirects=True) as client:
        info = client.get(
            "https://passport.bilibili.com/x/passport-login/web/cookie/info",
            params={"csrf": csrf},
            headers={"cookie": cookie_str},
        ).json()
        if info.get("code") != 0:
            raise RuntimeError(f"cookie/info code={info.get('code')} {info.get('message')}")
        if not info.get("data", {}).get("refresh") and not force:
            return {"refreshed": False, "reason": "cookie still fresh",
                    "cookie": cookie_str, "refresh_token": refresh_token}
        timestamp = info["data"]["timestamp"]

        path = _correspond_path(timestamp)
        html = client.get(
            f"https://www.bilibili.com/correspond/1/{path}",
            headers={"cookie": cookie_str},
        ).text
        match = re.search(r'<div id="1-name">\s*([0-9a-fA-F]+)\s*</div>', html)
        if not match:
            raise RuntimeError("refresh_csrf not found (correspondPath expired or cookie invalid)")
        refresh_csrf = match.group(1)

        refreshed = client.post(
            "https://passport.bilibili.com/x/passport-login/web/cookie/refresh",
            headers={"cookie": cookie_str, "content-type": "application/x-www-form-urlencoded"},
            data={"csrf": csrf, "refresh_csrf": refresh_csrf, "source": "main_web", "refresh_token": refresh_token},
        )
        rj = refreshed.json()
        if rj.get("code") != 0:
            raise RuntimeError(f"cookie/refresh code={rj.get('code')} {rj.get('message')}")
        new_token = rj.get("data", {}).get("refresh_token")
        if not new_token:
            raise RuntimeError("cookie/refresh returned no refresh_token")

        # Apply the new Set-Cookie values (SESSDATA, bili_jct, DedeUserID, sid, …).
        for name, value in refreshed.cookies.items():
            jar[name] = value
        new_cookie = _serialize(jar)
        new_csrf = jar.get("bili_jct")
        if not new_csrf:
            raise RuntimeError("refreshed cookie missing bili_jct")

        # Confirm with the NEW csrf + the OLD token to invalidate the old cookie.
        confirm = client.post(
            "https://passport.bilibili.com/x/passport-login/web/confirm/refresh",
            headers={"cookie": new_cookie, "content-type": "application/x-www-form-urlencoded"},
            data={"csrf": new_csrf, "refresh_token": refresh_token},
        ).json()
        if confirm.get("code") != 0:
            logger.warning("bili confirm/refresh non-zero: %s %s", confirm.get("code"), confirm.get("message"))

        return {"refreshed": True, "reason": "ok", "cookie": new_cookie, "refresh_token": new_token}

File: cf/worker/src/lib/biliRefresh.ts
// Bilibili web Cookie refresh.
//
// Bilibili's web cookies decay as "sensitive" endpoints are hit; the official
// mechanism is to periodically refresh them with a persistent refresh_token.
// Flow (see https://github.com/SocialSisterYi/bilibili-API-collect login docs):
//   1. GET  /x/passport-login/web/cookie/info?csrf=<bili_jct>  -> {refresh, timestamp}
//   2. correspondPath = hex(RSA-OAEP-SHA256(pubkey, "refresh_"+timestamp))
//      GET https://www.bilibili.com/correspond/1/<path>        -> #1-name = refresh_csrf
//   3. POST /x/passport-login/web/cookie/refresh               -> new Set-Cookie + new refresh_token
//   4. POST /x/passport-login/web/confirm/refresh (new csrf, OLD token) -> invalidate old cookie
// New cookie + refresh_token are persisted back to KV.
import type { Env } from "../env";
import { loadBiliAuth, saveBiliAuth } from "./biliAuth";

const UA =
  "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0 Safari/537.36";

// Bilibili's well-known RSA public key (JWK) for correspondPath generation.
const PUBKEY_JWK: JsonWebKey = {
  kty: "RSA",
  n: "y4HdjgJHBlbaBN04VERG4qNBIFHP6a3GozCl75AihQloSWCXC5HDNgyinEnhaQ_4-gaMud_GF50elYXLlCToR9se9Z8z433U3KjM-3Yx7ptKkmQNAMggQwAVKgq3zYAoidNEWuxpkY_mAitTSRLnsJW-NCTa0bqBFF6Wm1MxgfE",
  e: "AQAB",
};

export interface RefreshOutcome {
  refreshed: boolean;
  reason: string;
}

function parseCookie(s: string): Map<string, string> {
  const jar = new Map<string, string>();
  for (const part of (s || "").split(";")) {
    const t = part.trim();
    const i = t.indexOf("=");
    if (i <= 0) continue;
    jar.set(t.slice(0, i).trim(), t.slice(i + 1).trim());
  }
  return jar;
}

function serializeCookie(jar: Map<string, string>): string {
  return [...jar.entries()].map(([k, v]) => `${k}=${v}`).join("; ");
}

// Apply Set-Cookie response headers onto the jar (name=value, ignoring attrs).
function mergeSetCookies(jar: Map<string, string>, setCookies: string[]): void {
  for (const sc of setCookies) {
    const first = (sc.split(";", 1)[0] || "").trim();
    const i = first.indexOf("=");
    if (i <= 0) continue;
    const name = first.slice(0, i).trim();
    const value = first.slice(i + 1).trim();
    if (value && value.toLowerCase() !== "deleted") jar.set(name, value);
  }
}

async function correspondPath(timestamp: number): Promise<string> {
  const key = await crypto.subtle.importKey(
    "jwk",
    PUBKEY_JWK,
    { name: "RSA-OAEP", hash: "SHA-256" },
    false,
    ["encrypt"],
  );
  const data = new TextEncoder().encode(`refresh_${timestamp}`);
  const enc = new Uint8Array(await crypto.subtle.encrypt({ name: "RSA-OAEP" }, key, data));
  let hex = "";
  for (const b of enc) hex += b.toString(16).padStart(2, "0");
  return hex;
}

async function biliGetJson(url: string, cookie: string): Promise<Record<string, any>> {
  const res = await fetch(url, { headers: { "user-agent": UA, cookie } });
  return (await res.json()) as Record<string, any>;
}

// workerd supports Headers.getSetCookie() at runtime (one entry per Set-Cookie),
// but it isn't in @cloudflare/workers-types yet.
function getSetCookies(res: Response): string[] {
  const h = res.headers as unknown as { getSetCookie?: () => string[] };
  return h.getSetCookie ? h.getSetCookie() : [];
}

// Refresh the stored Bilibili cookie if Bilibili says it needs it (or force).
// Returns whether an actual refresh happened. Throws on a hard failure mid-flow
// (caller logs; the previous cookie stays in KV untouched).
export async function refreshBilibiliCookie(
  env: Env,
  opts: { force?: boolean } = {},
): Promise<RefreshOutcome> {
  const auth = await loadBiliAuth(env);
  if (!auth) return { refreshed: false, reason: "no stored bilibili auth" };
  if (!auth.refresh_token) return { refreshed: false, reason: "no refresh_token (set BILIBILI_REFRESH_TOKEN)" };

  const jar = parseCookie(auth.cookie);
  const csrf = jar.get("bili_jct");
  if (!csrf) return { refreshed: false, reason: "stored cookie missing bili_jct" };

  // 1. Does it need refreshing?
  const info = await biliGetJson(
    `https://passport.bilibili.com/x/passport-login/web/cookie/info?csrf=${csrf}`,
    auth.cookie,
  );
  if (info.code !== 0) return { refreshed: false, reason: `cookie/info code=${info.code} ${info.message ?? ""}` };
  const timestamp: number = info.data?.timestamp ?? Date.now();
  if (!info.data?.refresh && !opts.force) return { refreshed: false, reason: "cookie still fresh" };

  // 2. correspondPath -> refresh_csrf
  const path = await correspondPath(timestamp);
  const html = await (
    await fetch(`https://www.bilibili.com/correspond/1/${path}`, {
      headers: { "user-agent": UA, cookie: auth.cookie },
    })
  ).text();
  const match = html.match(/<div id="1-name">\s*([0-9a-fA-F]+)\s*<\/div>/);
  if (!match) throw new Error("refresh_csrf not found (correspondPath expired or cookie invalid)");
  const refreshCsrf = match[1];

  // 3. Refresh -> new Set-Cookie + new refresh_token. Keep the OLD token for step 4.
  const oldToken = auth.refresh_token;
  const refreshRes = await fetch("https://passport.bilibili.com/x/passport-login/web/cookie/refresh", {
    method: "POST",
    headers: { "user-agent": UA, cookie: auth.cookie, "content-type": "application/x-www-form-urlencoded" },
    body: new URLSearchParams({ csrf, refresh_csrf: refreshCsrf, source: "main_web", refresh_token: oldToken }).toString(),
  });
  const setCookies = getSetCookies(refreshRes);
  const refreshJson = (await refreshRes.json()) as Record<string, any>;
  if (refreshJson.code !== 0) {
    throw new Error(`cookie/refresh code=${refreshJson.code} ${refreshJson.message ?? ""}`);
  }
  const newToken: string | undefined = refreshJson.data?.refresh_token;
  if (!newToken) throw new Error("cookie/refresh returned no refresh_token");

  mergeSetCookies(jar, setCookies);
  const newCsrf = jar.get("bili_jct");
  if (!newCsrf) throw new Error("refreshed cookie missing bili_jct");
  const newCookie = serializeCookie(jar);

  // 4. Confirm with the NEW csrf and the OLD token (invalidates the old cookie).
  const confirmRes = await fetch("https://passport.bilibili.com/x/passport-login/web/confirm/refresh", {
    method: "POST",
    headers: { "user-agent": UA, cookie: newCookie, "content-type": "application/x-www-form-urlencoded" },
    body: new URLSearchParams({ csrf: newCsrf, refresh_token: oldToken }).toString(),
  });
  const confirmJson = (await confirmRes.json()) as Record<string, any>;
  if (confirmJson.code !== 0) {
    // The new cookie is already valid; failing to invalidate the old one is
    // non-fatal, so persist and warn rather than discarding the refresh.
    console.warn("bilibili confirm/refresh non-zero", confirmJson.code, confirmJson.message);
  }

  await saveBiliAuth(env, { cookie: newCookie, refresh_token: newToken, updated_at: new Date().toISOString() });
  return { refreshed: true, reason: "ok" };
}
